src/script/miniscript.h

```diff
@@ -746,6 +763,36 @@ struct Node {
         return {{}, {}};
     }
 
+    /** Check whether any key is repeated.
+     * This uses a custom key comparator provided by the context in order to still detect duplicates
+     * for more complicated types.
+     */
+    template<typename Ctx> bool ContainsDuplicateKey(const Ctx& ctx) const {
+        using set = std::set<Key, std::function<bool(const Key& a, const Key& b)>>;
```

Comment:
`std::function` may have allocation overhead, and results in dynamic binding (functions pointers rather than compile-time resolving). Here is an alternative, though it's not as elegant as I'd like:

```c++
--- a/src/script/miniscript.h
+++ b/src/script/miniscript.h
@@ -767,14 +767,20 @@ public:
      * for more complicated types.
      */
     template<typename Ctx> bool ContainsDuplicateKey(const Ctx& ctx) const {
-        using set = std::set<Key, std::function<bool(const Key& a, const Key& b)>>;
+        // We cannot use a lambda here, as lambdas are non assignable, and the set operations
+        // below require moving the comparators around.
+        struct Comp {
+            const Ctx* ctx_ptr;
+            Comp(const Ctx& ctx) : ctx_ptr(&ctx) {}
+            bool operator()(const Key& a, const Key& b) const { return ctx_ptr->KeyCompare(a, b); }
+        };
+        using set = std::set<Key, Comp>;
 
         auto upfn = [this, &ctx](const Node& node, Span<set> subs) -> std::optional<set> {
             if (&node != this && node.duplicate_key) return {};
 
             size_t keys_count = node.keys.size();
-            auto comp = [&ctx](const Key& a, const Key& b) -> bool { return ctx.KeyCompare(a, b); };
-            set key_set{node.keys.begin(), node.keys.end(), comp};
+            set key_set{node.keys.begin(), node.keys.end(), Comp(ctx)};
             if (key_set.size() != keys_count) return {};
 
             for (auto& sub: subs) {
```

Reply:
[TIL](https://stackoverflow.com/a/9088690/12967587), thanks. Done.